### ISSUE TITLE
Same matrix multiplication get different results on CPU and GPU

### ISSUE THREAD

**Issue by MosquitoStar** [pytorch/pytorch#116966]
### 🐛 Describe the bug

I run the following matrix multiplication code on my Linux server.

```python
a = torch.tensor([
    [ 8.3899e-01,  5.4414e-01,  3.3142e+05],
    [-5.4414e-01,  8.3899e-01,  4.6909e+06],
    [ 0.0000e+00,  0.0000e+00,  1.0000e+00]]).type(torch.float)
b = torch.tensor([
    [ 8.3899e-01, -5.4414e-01,  2.2745e+06],
    [ 5.4414e-01,  8.3899e-01, -4.1160e+06],
    [ 0.0000e+00,  0.0000e+00,  1.0000e+00]]).type(torch.float)

print(a @ b)
print(a.cuda() @ b.cuda())
print(a.double().cuda() @ b.double().cuda())
print(a.cuda().double() @ b.cuda().double())
```

And the terminal outputs are

```
tensor([[  1.0000,   0.0000,  22.5000],
        [  0.0000,   1.0000, -29.0000],
        [  0.0000,   0.0000,   1.0000]])
tensor([[ 9.9957e-01,  0.0000e+00,  1.0780e+03],
        [ 0.0000e+00,  9.9957e-01, -9.1400e+02],
        [ 0.0000e+00,  0.0000e+00,  1.0000e+00]], device='cuda:0')
tensor([[  1.0000,   0.0000,  22.5308],
        [  0.0000,   1.0000, -29.1165],
        [  0.0000,   0.0000,   1.0000]], device='cuda:0', dtype=torch.float64)
tensor([[  1.0000,   0.0000,  22.5308],
        [  0.0000,   1.0000, -29.1165],
        [  0.0000,   0.0000,   1.0000]], device='cuda:0', dtype=torch.float64)
```

It is strange that the second output tensor is very different from the others. I suspect there is a precision problem in cuda float tensor, but I could not tell. 

### Versions

The Linux system of my server is Ubuntu 20.04.4 LTS (GNU/Linux 5.4.0-91-generic x86_64). PyTorch version is 1.9.0+cu111. CUDA version is 11.7. The GPU is RTX A5000.

I also tried to reproduce this bug on my office computer but failed, the outputs are the same. The system on my office computer is Windows 10 Enterprise. PyTorch version is also 1.9.0+cu111. CUDA version is 11.1. The GPU is RTX 2080Ti

**Comment by tringwald** [comment]
Can you try again with `torch.backends.cuda.matmul.allow_tf32 = False`?

**Comment by razzzeeev** [comment]
CPUs and GPUs handle floating-point arithmetic differently. CPUs, especially in PyTorch's default settings, often use double precision (float64) for calculations, which provides more accuracy. GPUs, on the other hand, are optimized for single precision (float32) to gain performance benefits. This difference can lead to variance in results.

The difference in how results come out between single precision (float32) and double precision (float64) in GPU , it's 'cause of the way numbers get handled:

How Numbers are Stored: Float64 gives more bits to each part of a number (like the sign, exponent, and the significand) than float32. This means it can keep numbers more accurately.

Mistakes in Precision and Rounding Off: Since float64 uses more bits for the significand, it's got fewer and smaller errors when rounding off numbers. With float32, these little mistakes can add up over a bunch of calculations, making the final results quite different.

Dealing with Really Big or Small Numbers: Float64 handles super big or tiny numbers better because it's got a wider range for the exponent. This makes it more precise for calculations with these kinds of numbers.

How GPUs Work: GPUs are made to be better at float32 operations. They're quicker and don't need as many resources. That's why float32 is usually the go-to for stuff where being fast is more important than being super accurate.

So, all in all, while float32 is faster in crunching numbers, it's a bit less precise. On the other hand, float64 is more about accuracy but it's slower and needs more power to work. Which one you pick depends on what you need more in your work – speed or precision.

**Comment by MosquitoStar** [comment]
> Can you try again with `torch.backends.cuda.matmul.allow_tf32 = False`?

Thank you very much! The outputs are the same now.

**Comment by janeyx99** [comment]
Cool! Closing as this has been resolved. Please reopen if that's not the case.
